crates/red_knot_python_semantic/src/types.rs

```diff
@@ -657,17 +657,23 @@ impl<'db> Type<'db> {
         // TODO: Once we have support for final classes, we can establish that
         // `Type::SubclassOf('FinalClass')` is equivalent to `Type::ClassLiteral('FinalClass')`.
 
-        // TODO: The following is a workaround that is required to unify the two different
-        // versions of `NoneType` in typeshed. This should not be required anymore once we
-        // understand `sys.version_info` branches.
+        // TODO: The following is a workaround that is required to unify the two different versions
+        // of `NoneType` and `NoDefaultType` in typeshed. This should not be required anymore once
+        // we understand `sys.version_info` branches.
         self == other
             || matches!((self, other),
                 (
                     Type::Instance(InstanceType { class: self_class }),
                     Type::Instance(InstanceType { class: target_class })
                 )
-                if self_class.is_known(db, KnownClass::NoneType) &&
-                target_class.is_known(db, KnownClass::NoneType))
+                if (
+                    self_class.is_known(db, KnownClass::NoneType) &&
+                    target_class.is_known(db, KnownClass::NoneType)
+                ) || (
+                    self_class.is_known(db, KnownClass::NoDefaultType) &&
+                    target_class.is_known(db, KnownClass::NoDefaultType)
+                )
```

Comment:
I think this makes sense, because I'm guessing `NoDefaultType` sometimes comes from `typing` and sometimes comes from `typing_extensions`. But it doesn't actually seem to be tested anywhere; no tests fail if I revert this change.

I also wonder if this and some code in `display.rs` could be micro-optimised slightly by reducing some Salsa lookups:

```diff
--- a/crates/red_knot_python_semantic/src/types.rs
+++ b/crates/red_knot_python_semantic/src/types.rs
@@ -666,13 +666,11 @@ impl<'db> Type<'db> {
                     Type::Instance(InstanceType { class: self_class }),
                     Type::Instance(InstanceType { class: target_class })
                 )
-                if (
-                    self_class.is_known(db, KnownClass::NoneType) &&
-                    target_class.is_known(db, KnownClass::NoneType)
-                ) || (
-                    self_class.is_known(db, KnownClass::NoDefaultType) &&
-                    target_class.is_known(db, KnownClass::NoDefaultType)
-                )
+                if {
+                    let self_class = self_class.known(db);
+                    matches!(self_class, Some(KnownClass::NoDefaultType | KnownClass::NoneType))
+                        && self_class == target_class.known(db)
+                }
             )
     }
 
diff --git a/crates/red_knot_python_semantic/src/types/display.rs b/crates/red_knot_python_semantic/src/types/display.rs
index 5bcafd902..bb0603ed0 100644
--- a/crates/red_knot_python_semantic/src/types/display.rs
+++ b/crates/red_knot_python_semantic/src/types/display.rs
@@ -66,15 +66,13 @@ impl Display for DisplayRepresentation<'_> {
             Type::Any => f.write_str("Any"),
             Type::Never => f.write_str("Never"),
             Type::Unknown => f.write_str("Unknown"),
-            Type::Instance(InstanceType { class })
-                if class.is_known(self.db, KnownClass::NoneType) =>
-            {
-                f.write_str("None")
-            }
-            Type::Instance(InstanceType { class })
-                if class.is_known(self.db, KnownClass::NoDefaultType) =>
-            {
-                f.write_str("NoDefault")
+            Type::Instance(InstanceType { class }) => {
+                let representation = match class.known(self.db) {
+                    Some(KnownClass::NoDefaultType) => "NoDefault",
+                    Some(KnownClass::NoneType) => "None",
+                    _ => class.name(self.db),
+                };
+                f.write_str(representation)
             }
             // `[Type::Todo]`'s display should be explicit that is not a valid display of
             // any other type
@@ -87,7 +85,6 @@ impl Display for DisplayRepresentation<'_> {
             Type::SubclassOf(SubclassOfType { class }) => {
                 write!(f, "type[{}]", class.name(self.db))
             }
-            Type::Instance(InstanceType { class }) => f.write_str(class.name(self.db)),
             Type::KnownInstance(known_instance) => f.write_str(known_instance.as_str()),
             Type::FunctionLiteral(function) => f.write_str(function.name(self.db)),
             Type::Union(union) => union.display(self.db).fmt(f),
```

Reply:
https://github.com/astral-sh/ruff/pull/14232